### PR TITLE
Add new rule: `no-yield-to-default`

### LINKED ISSUE #1352: Rule-request: no yield to default
```hbs
{{yield ... to="default"}} -> {{yield ...}}
{{has-block "default"}} -> {{has-block}}
{{has-block-params "default"}} -> {{has-block-params}}
{{hasBlock "default"}} -> {{hasBlock}}
{{hasBlockParams "default"}} -> {{hasBlockParams}}
```

### PR DESCRIPTION
**Pull request by rajasegar**
Fixes #1352

### CODE REVIEW

File: docs/rule/no-yield-to-default.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
+# no-yield-to-default
```
Comment (rwjblue) [inline docs/rule/no-yield-to-default.md:1]: Still a bunch of `TODO`s in this file.

File: lib/rules/no-yield-to-default.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+const ERROR_MESSAGE = 'You should not yield to default';
+
+const DISALLOWED_PATHS = ['yield', 'has-block', 'has-block-params', 'hasBlock', 'hasBlockParams'];
+
+module.exports = class NoYieldToDefault extends Rule {
+  visitor() {
+    return {
+      MustacheStatement(node) {
+        if (DISALLOWED_PATHS.includes(node.path.original)) {
```
Comment (rwjblue) [inline lib/rules/no-yield-to-default.js:13]: We can't assume that `.path.original` is present. For example with `{{"foo"}}` there is no `.path.original` (I think its `.path.value`?).

File: lib/rules/no-yield-to-default.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+          let isDefaultHashPair = node.hash.pairs.some((p) => {
+            return p.key === 'to' && p.value.value === 'default';
```
Comment (rwjblue) [inline lib/rules/no-yield-to-default.js:15]: Checking `.value.value` will error if the hash pair value is not a static string.

File: test/unit/rules/no-yield-to-default-test.js [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+  name: 'no-yield-to-default',
+
+  config: true,
+
+  good: [
+    '{{yield}}',
+    '{{has-block}}',
+    '{{has-block-params}}',
+    '{{hasBlock}}',
+    '{{hasBlockParams}}',
```
Comment (rwjblue) [inline test/unit/rules/no-yield-to-default-test.js:17]: Can you add a few more "good" scenarios for other expected invocations of these keywords (e.g. passing the expected block names)?

File: test/unit/rules/no-yield-to-default-test.js [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+  ],
+
+  bad: [
```
Comment (rwjblue) [inline test/unit/rules/no-yield-to-default-test.js:20]: Can you add more bad examples for (permutations of each of the disallowed keywords):

```hbs
{{#if (has-block "default")}}

{{/if}}
```

```hbs
{{if (has-block "default") "some-value"}}
```

Comment (rajasegar) [reply test/unit/rules/no-yield-to-default-test.js:25]: done

File: test/unit/rules/no-yield-to-default-test.js (hunk 4 above)
Comment (rajasegar) [reply test/unit/rules/no-yield-to-default-test.js]: done

File: lib/rules/no-yield-to-default.js (hunk 2 above)
Comment (rajasegar) [reply lib/rules/no-yield-to-default.js:13]: But we are looking for `{{yield}}` and `{{has-block}}` nodes right, for them the `path.original` is always there.

File: lib/rules/no-yield-to-default.js (hunk 3 above)
Comment (rajasegar) [reply lib/rules/no-yield-to-default.js:15]: Which means we should look out for dynamic values for named blocks like `to` ?

File: docs/rule/no-yield-to-default.md (hunk 1 above)
Comment (rajasegar) [reply docs/rule/no-yield-to-default.md:1]: Removed. Anyway the docs require a better description and references I think.

Comment (rwjblue) [reply docs/rule/no-yield-to-default.md:1]: @chancancode - Can you help with some prose/info here?

Comment (chancancode) [reply docs/rule/no-yield-to-default.md:1]: Something like:

> The `yield` keyword can be used for invoking blocks passed into a component. The `to` named argument specifies which of the blocks to yield too. Specifying `{{yield to="default"}}` is unnecessary, as that is the default behavior. Likewise, `{{has-block}}` and `{{has-block-params}}` also defaults to checking the "default" block.

File: lib/rules/no-yield-to-default.js (hunk 3 above)
Comment (chancancode) [reply lib/rules/no-yield-to-default.js:15]: That is in fact an error, but I think for the purpose of this rule you can just ignore that cause (don't blow up) and let something else handle it. (IIRC, the glimmer compiler already has a syntax error for that.)

File: docs/rules.md [hunk 6]
```diff
@@ -58,6 +58,7 @@
 * [no-whitespace-for-layout](rule/no-whitespace-for-layout.md)
 * [no-whitespace-within-word](rule/no-whitespace-within-word.md)
 * [no-yield-only](rule/no-yield-only.md)
+* [no-yield-to-default](rule/no-yield-to-default.md)
```
Comment (bmish) [inline docs/rules.md:61]: This file has been deleted, please rebase on the latest master and run `yarn update` to update the README.

**Comment by bmish** [comment]
@rajasegar is this ready?

**Comment by rajasegar** [comment]
No @bmish not yet, @rwjblue had some concerns regarding the use of `.value.value` I am still wondering we should handle this for dynamic values..

**Comment by bmish** [comment]
@rajasegar I see, well heads up that we will have a major release soon, so it's a good time to get any new rules ready and merged.

File: lib/rules/no-yield-to-default.js (hunk 2 above)
Comment (rwjblue) [reply lib/rules/no-yield-to-default.js]: > But we are looking for {{yield}} and {{has-block}} nodes right, for them the path.original is always there.

This is no longer correct, `node.path` is allowed to be roughly "anything". For example, in Ember 3.25+ this is perfectly valid:

```hbs
{{(call-some-helper-that-returns-a-helper) whatever-here="lol"}}
```

**Pull request merged**